docs/design/loadpath.md

```diff
@@ -0,0 +1,284 @@
+be handled by a user-space process that has the required capabilities, and then
+the result of the syscall is returned to the caller.
+
+When presented as an alternative to the BPF Token patchset, it was dismissed
+since Christian Brauner (the author of the blog post) described it as
+"impractical". However, there is a project called [Seitan] which seeks to
+address some of the practicality issues, and also provide container runtime
+integration.
+
+The integration would look something like this:
+
+1. bpfman might be able to provide information to Seitan that can be used to
+   generate the seccomp filters that are required to proxy syscalls.
+1. bpfman might do nothing, except for preparing a bpffs filesystem for the
+   container runtime to use.
+
+[The Seccomp Notifier: New Frontiers in Unprivileged Container Development]: https://people.kernel.org/brauner/the-seccomp-notifier-new-frontiers-in-unprivileged-container-development
+[Seitan](https://seitan.rocks/seitan/about/)
+
+#### Policing CAP_BPF (via LSM or otherwise)
+
+It's possible to police how the bpf syscall is used by a process using LSM.
+
+This would likely require some "profiling" of an eBPF application in order
+to derive a policy that we could apply - using eBPF LSM or something else.
+This would be a one-time cost at development time.
+You could choose to issue warnings or block operations that are not permitted by
+the policy.
+
+Such an approach was [demonstrated at CloudNativeSecurityCon] by the Tetragon
+team.
+
+The downside here is that the security guarantee is only ever as strong as your
+policy. The policies required may end up being pretty verbose and complex.
+
+[demonstrated at CloudNativeSecurityCon](https://youtu.be/UBVTJ0LeXxc?si=lGjcHK_CAuYd1d5v)
+
+#### CAP_BPF Conclusion
+
+The use of eBPF in rootless containers (user namespaces) is still
+a work in progress. In the short term, the following work items are required:
+
+1. Teach bpfman to generate BPF tokens.
+1. Permit BPF token generation into the bpfman CSI plugin.
+1. Try loading eBPF programs from a user namespace using BPF tokens and confirm
+   which use-cases are supported, and which are not.
+1. Present the results of that investigation to the kernel community and
+   determine whether it's possible to support the remaining use-cases with
+   BPF tokens, or whether they will always be required to be run as root.
+
+In the medium term, we should consider the following:
+
+1. Work with the Seitan authors on supporting the bpf syscall.
+1. Scope out what controls Seitan offers over which syscalls can be proxied,
+   and how much room we have to define policy there.
+1. See what value, if any, bpfman could add on top of this.
+
+Longer term we might consider working with the Tetragon team on policy
+profiling and enforcement for eBPF programs.
+
+## In Summary
+
+In summary, we have a path forward to remove bpfman from the load path,
+while still providing some of the benefits of our current design.
```

Comment:
From reading the above it seems that the proposal is that bpfman can operate in one of two modes:
- The current "load on behalf of the container" mode
- A mode where bpfman becomes the entity that does the user namespace hole punching required to grant an entity CAP_BPF inside a container

As noted above, I'm not really sure the fsverity stuff adds anything that just signing the container image (combined with granting it CAP_BPF directly or by token) doesn't already provide. So, like, if the proposal is "we'll try to make people happy by making bpfman the system level daemon responsible for the privilege hole punching", that's fine, but a lot of the details could be omitted in that case? :)

Reply:
Yes agreed, this did get pretty wordy in the end.
And your summary is pretty accurate.

I felt that we still needed to mention signing as that's one of the features of bpfman "load-on-behalf-of-container" mode. While we can hole punch, we'd rather not lose a security feature, so having another signing solution would be great.

I'll re-organize this and move a lot of the wordier bits into appendicies.

Thanks 🙏